efivar: new package

oskarirauta opened the pull request:
Maintainer: me / @oskarirauta 
Compile tested: x86-64, recent git
Run tested: x86-64, recent git

Description:

efivar library. Required by my another PR #20277 (efibootmgr)

Signed-off-by: Oskari Rauta <oskari.rauta@gmail.com>

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,67 @@
+PKG_SOURCE_URL:=https://github.com/rhboot/efivar/releases/download/$(PKG_VERSION)
+PKG_HASH:=f018ed6e49c5f1c16d336d9fd7687ce87023276591921db1e49a314ad6515349
+
+PKG_LICENSE:=LGPL-2.1-only
+PKG_LICENSE_FILES:=COPYING
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+PKG_MAINTAINER:=OSkari Rauta <oskari.rauta@gmail.com>
```
```suggestion
PKG_LICENSE:=LGPL-2.1-only
PKG_LICENSE_FILES:=COPYING
PKG_MAINTAINER:=OSkari Rauta <oskari.rauta@gmail.com>

PKG_BUILD_PARALLEL:=1
PKG_INSTALL:=1
```

1715173329 commented on libs/efivar/files/efivar-guids.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef EFIVAR_GUIDS_H
```
where do these files come from? can't them be upstreamed?

oskarirauta replied:
normal linux efivar headers....

oskarirauta replied:
They came from somewhere - but I cannot find them now afterwards from anywhere. And I though that they were needed for a reason.. Maybe they were for older version of efivar, hard to say. I removed them and patches equipping them.

1715173329 commented on libs/efivar/patches/001-workarounds.patch:
```diff
@@ -0,0 +1,28 @@
+ 
+ LD_FLAVOR := $(shell $(LD) --version | grep -E '^(LLD|GNU ld)'|sed 's/ .*//g')
+ LD_VERSION := $(shell $(LD) --version | grep -E '^(LLD|GNU ld)'|sed 's/.* //')
+-# I haven't tested 2.36 here; 2.35 is definitely broken and 2.37 seems to work
```
we're at 2.39 already, is it still required?

oskarirauta replied:
You are correct :)

oskarirauta replied:
Worked on my end without patch. I removed it.

1715173329 commented on libs/efivar/patches/002-musl-compat.patch:
```diff
@@ -0,0 +1,187 @@
+--- a/src/compiler.h
```
I found this patch at https://git.alpinelinux.org/aports/tree/main/efivar/musl-compat.patch.
Please keep his patch header here.

oskarirauta replied:
Sure. Header was lost during patch refreshing for quilt compatibility (dirty patches...)

oskarirauta commented:
It says that 1 change is requested, but I can't find it from anywhere.... What is it?

1715173329 commented:
> It says that 1 change is requested, but I can't find it from anywhere.... What is it?

It just meant somebody left a review.

oskarirauta commented:
> > It says that 1 change is requested, but I can't find it from anywhere.... What is it?
> 
> It just meant somebody left a review.

Sometimes, when I don't press "Resolve" **before** pushing commit, platform seems to remember that there's a request for changes, but it becomes inaccessible since content has changed, it's propably that then...

1715173329 commented:
Nah.

oskarirauta commented:
Could you verify if this is ok. I accidentally seem to have removed request for neheb's review.

@neheb - please, review.

oskarirauta commented:
I can't remember where I got those files- but seems that more or less they are needed. Problem with finding out where I got those headers, is that I created this package for personal use ages ago, had difficulties to build it and found out that I needed files that I included- but source... Well, it seems to have disappeared..

It is rather difficult to "test" this locally, as I build on a system that is compatible with target (alpine x86-64)- so most built software can be executed on a host- then there occurs issues when I push the commit, as build environment is different. Sorry about that, I'll try to get this to build properly again.

oskarirauta commented:
Ok, now it should be able to build again.

@1715173329 now you might want to ask again where did these files come from- answer is that there is util included: makeguids - which generates these files. Ofcourse it will be cross-compiled, so there is no way to use it on the host system, except in my case since my host system has no trouble to execute it. So, I generated those files and added them manually and removed generation of files from Makefile with patch 007.

missing qsort_r patch and files- I did not add them, as they are not necessary (anymore), like I said; I first added this package some time ago, and at that time, musl did not have qsort_r, so I found a patch and these files from somewhere, and then added them; now we run with newer musl, which has qsort_r, so that problem is in the past. Propably these files are still in somewhere, but google can't find them- I just found other issues with similar/same solution conversations on some github projects, which revealed history behind the solution.

oskarirauta commented:
Okay, now it's fixed. I'll rebase and then this is ready for reviewing.

1715173329 commented:
> Ofcourse it will be cross-compiled, so there is no way to use it on the host system, except in my case since my host system has no trouble to execute it.

https://github.com/rhboot/efivar/blob/1753149d4176ebfb2b135ac0aaf79340bf0e7a93/src/Makefile#L62-L69
it calls host gcc to build - in theory it should work.

oskarirauta commented:
> > Ofcourse it will be cross-compiled, so there is no way to use it on the host system, except in my case since my host system has no trouble to execute it.
> 
> https://github.com/rhboot/efivar/blob/1753149d4176ebfb2b135ac0aaf79340bf0e7a93/src/Makefile#L62-L69 it calls host gcc to build - in theory it should work.

Yes. In theory. Still in real life it failed on checks here, because it was not possible to execute makeguids.

oskarirauta commented:
@dangowrt please, take a look at this you too. As you were more familiar with EFI.

oskarirauta commented:
@1715173329 
@dangowrt 

What would it take to get this merged?

dangowrt reviewed (changes requested):
Imho it's better to generate host-builds for what ever is needed to generate `efivar-guids.h` from source instead of carrying a built version in the tree. Ie. add a host-build like it is done for e.g. for postgresql.

oskarirauta commented:
@dangowrt

> Imho it's better to generate host-builds for what ever is needed to generate `efivar-guids.h` from source instead of carrying a built version in the tree.

ccdf3c6f18475a48b1aab08f7b9585d15f0064bc should do it..

oskarirauta commented:
@dangowrt 

I have updated accordingly to your request. Would you review this again with these changes made?

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,88 @@
+
+PKG_LICENSE:=LGPL-2.1-only
+PKG_LICENSE_FILES:=COPYING
+PKG_BUILD_PARALLEL:=1
+PKG_BUILD_DEPENDS:=efivar/host
+PKG_INSTALL:=1
```
```suggestion

PKG_BUILD_PARALLEL:=1
PKG_BUILD_DEPENDS:=efivar/host
PKG_INSTALL:=1
```

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,88 @@
+PKG_BUILD_DEPENDS:=efivar/host
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/host-build.mk
```
```suggestion
```
just need to pass correct flags, host build can be removed.

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,88 @@
+  Tools and libraries to work with EFI variables
+endef
+
+HOST_MAKE_VARS += ERRORS=
+MAKE_VARS += \
+	MAKEGUIDS_BIN="$(STAGING_DIR_HOSTPKG)/usr/bin/efivar-makeguids" \
+	ERRORS=
```
```suggestion
MAKE_VARS += \
	ERRORS= \
	HOSTCC="$(HOSTCC)" \
	HOST_CFLAGS="$(HOST_CFLAGS)" \
	HOST_LDFLAGS="$(HOST_LDFLAGS)" \
	LIBDIR="/usr/lib"
```

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,88 @@
+endef
+
+define Package/efivar/install
+	$(INSTALL_DIR) $(1)/usr/bin $(1)/usr/lib/
+	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/{efisecdb,efivar} $(1)/usr/bin/
+	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/lib/{libefiboot.so.1.38,libefisec.so.1.38,libefivar.so.1.38} $(1)/usr/lib/
+	$(LN) libefiboot.so.1.38 $(1)/usr/lib/libefiboot.so
+	$(LN) libefiboot.so.1.38 $(1)/usr/lib/libefiboot.so.1
+	$(LN) libefisec.so.1.38 $(1)/usr/lib/libefisec.so
+	$(LN) libefisec.so.1.38 $(1)/usr/lib/libefisec.so.1
+	$(LN) libefivar.so.1.38 $(1)/usr/lib/libefivar.so
+	$(LN) libefivar.so.1.38 $(1)/usr/lib/libefivar.so.1
```
```suggestion
	$(INSTALL_DIR) $(1)/usr/bin
	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/efisecdb $(1)/usr/bin/
	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/efivar $(1)/usr/bin/

	$(INSTALL_DIR) $(1)/usr/lib
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libefiboot.so* $(1)/usr/lib/
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libefisec.so* $(1)/usr/lib/
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libefivar.so* $(1)/usr/lib/
```

1715173329 commented on libs/efivar/patches/006-fix-lib-path.patch:
```diff
@@ -0,0 +1,10 @@
+ PREFIX	?= /usr
+ EXEC_PREFIX ?= $(PREFIX)
+-LIBDIR	?= $(PREFIX)/lib64
++LIBDIR	?= $(PREFIX)/lib
```
replaced by `LIBDIR="/usr/lib"`. can be removed.

oskarirauta replied:
patch removed

1715173329 commented on libs/efivar/patches/007-use-host-makeguids.patch:
```diff
@@ -0,0 +1,19 @@
+ MAKEGUIDS_SOURCES = makeguids.c util.c
+ MAKEGUIDS_OBJECTS = $(patsubst %.S,%.o,$(patsubst %.c,%.o,$(MAKEGUIDS_SOURCES)))
+ MAKEGUIDS_OUTPUT = $(GENERATED_SOURCES) guids.lds
++MAKEGUIDS_BIN ?= ./makeguids
```
replaced by `HOSTCC="$(HOSTCC)" \`. can be removed.

oskarirauta replied:
patch removed

oskarirauta commented:
@1715173329 

That indeed was better solution than building a host package.

1715173329 commented on libs/efivar/Makefile:
```diff
@@ -0,0 +1,67 @@
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libefivar.so* $(1)/usr/lib/
+endef
+
+$(eval $(call HostBuild))
```
This should be removed too ;)

oskarirauta replied:
True. I'll remove that.. Just a second.

1715173329 reviewed (approved):
LGTM - Thank you!

oskarirauta commented:
Okay, that should be all. Thanks @1715173329 - You really did put some effort it, it's now very nice and clean :)
